app/models/source_resolver.rb

```diff
@@ -0,0 +1,34 @@
+class SourceResolver
+  def self.resolve(new_source)
+    new(new_source).resolve
+  end
+
+  def initialize(new_source)
+    @new_source = new_source
+  end
+
+  def resolve
+    return if @new_source.save
+    adjust_ranks
```

Comment:
I don't know what could make this fail. I imagined having to clean it up though when it did. I'm good if you're good.